File: Java/GenericStructures/Queue.java
package GenericStructures;

public class Queue<T> {
    private class Node{
        T data;
        Node next;
        Node(T data){
            this.data = data;
            this.next = null;
        }
    }
    private Node front, rear;
    private int size;

    Queue(){
        this.size = 0;
        this.front = null;
        this.rear = null;
    }

    public void enqueue(T data){
        Node node = new Node(data);

        if(is_empty()){
            this.front = node;
            this.rear = node;
        } else{
            node.next = this.rear;
            this.rear = node;
        }
        ++ this.size;
    }

    public void dequeue(){
        try {
            if(is_empty()) throw new Exception();
            // Node temp_node = this.front; // you can not use this?
            this.front = this.front.next;
            // temp_node = null; // you can not use this?
            
            -- this.size;
        } catch (Exception e) {
            System.out.print("Error. Queue is already empty.");
            return;
        }
    }

    public boolean is_empty(){
        if(size == 0 && this.front == null) return true;
        return false;
    }

    public T front(){
        return this.front.data;
    }
    
    public T rear(){
        return this.rear.data;
    }

    public int size(){return this.size;}
}

class MainQueue{
    public static void main(String[] args){

    }
}

File: Python/GenericStructures/Queue.py
from typing import TypeVar, Generic

T = TypeVar("T")


class Node(Generic[T]):
    next: Node[T] = None

    def __init__(self, data: T) -> None:
        self.data = data


class Queue(Generic[T]):
    size: int = 0
    front: Node[T] = None
    back: Node[T] = None

    def enqueue(self, data: int) -> None:
        node = Node[T](data)

        if self.is_empty():
            self.front = node
            self.front = node
        else:
            ptr = self.front
            while ptr.next != None:
                ptr = ptr.next
            ptr.next = node
            self.back = node

        self.size += 1

    def dequeue(self) -> None:
        if self.front == None:
            raise Exception("Error: Queue is already empty.")
        self.front = self.front.next
        self.size -= 1

    def display(self) -> None:
        ptr = self.front
        while ptr.next != None:
            print(ptr.data)
            ptr = ptr.next

    def is_empty(self) -> bool:
        return self.front == None and self.size == 0


if __name__ == "__main__":
    q = Queue()
    for i in range(0, 9):
        q.enqueue(i)

    # q.dequeue()
    q.display()
